feat: embedded superset dashboards

HendrikThePendric opened the pull request:
### TODO list based on KFMT session

(remove this section once completed)

- [x] Documentation added
- [x] Fix: The warning text in the delete dialog accepts focus which should not be possible
- [x] Fix: the delete button in the delete dialog should auto-focus
- [x] Fix: modal should get scrolling content in case of limited screen height. 
- [x] Fix: Stop disabling the “save” button in the create/update modal and run validation before taking action, so that the user gets better feedback of when the UUID is not valid.
- [x] Show improved error messages and prevent flashing of embedded superset error page where possible. To do this we will request a guest token once before we embed the dashboard and if this fails we do not attempt to embed the dashboard but show an error notification instead. This task is currently blocked. The `superset-gateway/api/guestTokens` endpoint should start returning HTTP errors and error codes first.
- [x] In the sharing dialog we should prevent the second tab to show, since it is not possible to apply sharing settings to embedded dashboard items.
- [x] [STATUS: ✅  DONE] Change "External data" to "External source"
- [x] [STATUS: ✅  DONE] Change "Embed a dashboard from a third-party source, like Superset." to "Embed a superset dashboard"
- [x] [STATUS: ⛔ NOT POSSIBLE] Attempt to fix fullscreen option of individual embedded items by adjusting sandboxing properties
- [x] [STATUS: ⛔ NOT POSSIBLE] Attempt to fix "share by email" option of embedded item by adjusting sandboxing properties

---

Implements [DHIS2-17891](https://dhis2.atlassian.net/browse/DHIS2-17891)

### References

- [Embedded Dashboards](https://docs.google.com/document/d/13435SfkE8UQqWYkOiPtKlGMot0METPTuAfooZjDUR5k) entry point
- [Design document](https://docs.google.com/document/d/1POHfTL48VfEN7VyR7yxIhg4TD08FI4YQVzuwciF_mb8)
- [Workflow in the Dashboards app](https://docs.google.com/document/d/1XBAll5QaNjqFEzs7IRaXK169JrmnWAXKnpBK9ChDKi0/edit?tab=t.0#heading=h.kzgtwuiq321s)
- [Superset embedded SDK](https://www.npmjs.com/package/@superset-ui/embedded-sdk)
- [Jira epic](https://dhis2.atlassian.net/browse/DHIS2-18728)

---

### Key features

1. When the app loads it checks if the backend supports Superset embedded dashboards and enables them if this is the case
2. When creating a new dashboard, the user can choose to create an internal or external dashboard
3. Title, Code, Description and Superset Embed ID can be provided when creating/editing a Superset Embedded dashboard, as well as the options to "Show chart controls on dashboard items" and/or "Expand filters" in the emedded dashboard (using the Superset Embed SDK)
4. Embedded Superset dashboards can be starred, shared, and the description can be displayed/hidden, just like internal dashboards
5. They cannot, however, be filtered, taken offline, go into slideshow mode, or be printed. So these menu items are disabled

---

### Partial PRs

Several PRs have been merged into this main feature PR. Please review PR descriptions and commit messages for a more granular view on the changes:
- #3181
- #3185
- #3192
- #3193
- #3197
- #3206


---

### Note for testers

Since this feature is only available for instances which have Superset included, you will need to use such an instance when testing, i.e. https://debug.dhis2.org/dev. To test the dashboard-app, you can either use the Netlify link or run the app locally.

---

### Screenshots

_Choosing a dashboard type_
<img width="1119" alt="Screenshot 2025-02-10 at 12 47 49" src="https://github.com/user-attachments/assets/9eeacd2d-34f1-444f-9279-ba1ca912f820" />


_Configuring a Superset embedded dashboard_
<img width="1122" alt="Screenshot 2025-02-10 at 12 48 26" src="https://github.com/user-attachments/assets/656c654f-c41c-4281-bda1-e7371972bc61" />


_Viewing a Superset embedded dashboard_
<img width="1120" alt="Screenshot 2025-02-10 at 12 49 00" src="https://github.com/user-attachments/assets/b0b4710e-bcd5-49f9-82a9-c82fe6011a20" />


[DHIS2-17891]: https://dhis2.atlassian.net/browse/DHIS2-17891?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

HendrikThePendric committed:
feat: fetch superset base url with system settings (#3181)

* feat: fetch superset base url with system settings

* fix: add hook dependencies and ensure fetchSuperSetBaseUrl is a stable reference

* feat: add hooks to read global superset data

* chore: ensure set in superset is in lower case

* chore: adjust request error message

HendrikThePendric committed:
feat: create embedded superset dashboard (#3185)

* chore: ignore aider ai files

* refactor: extract CreateDashboardButton and NavigationMenuDropdownButton components

* fix: read system settings from nested object in hooks

* feat: creation flow with conditional support for superset embedded dashboards

* chore: remove unused import

* feat: add form that can be used for creating and editing superset embedded dashboards

* chore: add file extensions to imports

* chore: addjust formatting

* feat: create superset embedded dashboards

* fix: complete hook dependencies

* fix: consolidate keyboard navigation implementation across modal forms

* fix: adjust modal form, radio card styles

* chore: add reminder to add tests for UUID validation

* fix: set focus on first fiel in modal form instead of setting focus to fieldset

* fix: align radio card outline visibility with other inputs

* chore: reorganise files

---------

Co-authored-by: Joseph John Aas Cooper <33054985+cooper-joe@users.noreply.github.com>
# Conflicts:
#	i18n/en.pot

HendrikThePendric committed:
feat: show embedded superset dashboard in iframe (#3192)

* chore: include embedded field when fetching dashboards

* chore: include embedded field when fetching an individual dashboard

* fix: ensure embedded property is on the selected dashboard state

* feat: show embedded-superset-dashboard and allow it to take up the full page

* chore: add @superset/embedded-sdk dependency

* chore: add a redux selector that returns the superset embed data

* chore: add an api helper that returns a superset guest token

* feat: load superset embedded dashboard into DOM

* fix: remove whitespace from dashboard container when displaying an embedded dashboard

* fix: adjust dashboard-type-radio focus style

* chore: set focus with focusVisible to theoretically show the blue focus outline

* feat: loading and error state on superset embedded dashboard

* fix: add padding to description when dasboard is embedded

* chore: fix failing tests
# Conflicts:
#	i18n/en.pot

HendrikThePendric committed:
feat: adjust dashboards bar for external dashboards (#3193)

* feat: show external data tag in dashboards-bar for embedded dashboards

* fix: disable unavailable buttons when showing embedded dashboard

* chore: remove unused import

* chore: fix broken unit test

* feat: edit and delete embedded dashboards using a modal

* refactor: encapsulate superset dashboard mutation logic in a hook

* chore: remove redundant use of nsSeparator in i18n.t call
# Conflicts:
#	i18n/en.pot

HendrikThePendric committed:
test: add e2e tests for superset embedded dashboards feature (#3197)

* chore: add excludeByVersionTag to cypress config and add tagify dependency

* chore: allow mixing cucumber and non-cucumber specs

* chore: temporarily disable all cypress-cucumber stuff

* test: add cypress test for creating a superset embedded dashboard

* chore: update dependencies so that cypress-tags could run

* chore: disable exclude by version tags and enable cucumber preprocessor

* chore: detect version during run and skip tests if below 42

* chore: keep cucumber specs separate from js specs to prevent

This prevents the js spec from being included in each cucumber feature.
https://github.com/badeball/cypress-cucumber-preprocessor/issues/1262

* fix: close more-menu after toggling dashboard starred state

* fix: ensure dashboard iframe is reloaded when dashboard embed data is updated

* test: finalize e2e test suite for superset embedded dashboard feature

* chore: ensure the before callback has a this scope

* chore: update yarn.lock

* chore: upgrade node version in all workflows

Ensure it  matches my local version where all tests pass

* chore: adjust cypress spec path

* chore: adjust test matrix so it includes both cucumber and regular cypress specs
# Conflicts:
#	cypress/e2e_cucumber/common/the_dashboard_description_is_(not)_displayed.js
#	cypress/e2e_cucumber/edit_dashboard/show_description.js
#	cypress/e2e_cucumber/view_dashboard/toggle_show_more_dashboards.js
#	cypress/e2e_cucumber/view_errors/error_while_show_description.js

HendrikThePendric committed:
chore: add unit tests and implement feedback (#3206)

* test: add unit tests for superset embedded fields state hook

* fix: use regular radio inputs in choose-dashboard-type modal

* fix: apply cursor style to outer container of radio

* fix: ensure selected is an object in edit dashboard unit test

* fix: adjust embedded dashboard fields

Since toggling the filter visibility is not working via the SDK
we have removed that option. And we have replaced it with an
option to expand/collapse the filters instead.

* chore: remove unused import

* fix: adjust buttons texts to design specs

* fix: make choose-dashboard-type modal dismissable

* chore: upgrade reselect

* fix: use memoized selector for embedData to avoid unnecessary dashboard reloads

* refactor: use ms prefix for reselect memoized selectors

* fix: adjust import name

* fix: add padding below description for embedded dashboards

While the dashboards themselves have some whitespace above
the filter controls section does not, so some whitespace
below the description is needed to avoid it touching the
filter controls.
# Conflicts:
#	i18n/en.pot

HendrikThePendric committed:
chore: fix problem in en.pot file

HendrikThePendric committed:
chore: always use "continue" text in choose-dashboard-type-modal button

jenniferarnesen commented on src/api/supersetGateway.js:
```diff
@@ -0,0 +1,57 @@
+import { useConfig } from '@dhis2/app-service-config'
+import { useCallback, useMemo } from 'react'
+
+const useSupersetApiUrl = (resource) => {
```
Perhaps each of these hooks could use a comment block with a little explanation?

jenniferarnesen commented on src/components/DashboardsBar/NavigationMenu/NavigationMenuDropdownButton.js:
```diff
@@ -0,0 +1,28 @@
+import i18n from '@dhis2/d2-i18n'
```
I was a little confused about what this component was. I guess because I don't think of it as "navigating" to a dashboard. I think of this component more as a "DashboardSelector"

HendrikThePendric replied:
But to get to another dashboard you change the route by pushing to the `history`...
> It exposes useful methods and properties that let you **navigate** back and forth through the user's history

_(Source: [MDN docs on the History API](https://developer.mozilla.org/en-US/docs/Web/API/History_API), emphasis mine)_

I'm not really convinced that "selection" is more appropriate

jenniferarnesen replied:
Fair point. But it's still a little misleading because there are other navigation elements. The "Edit" button, "New" button and "Print" menus. Maybe it could be called DashboardNavigator or DashboardNavigationMenuButton or similar. (Hard for me to say that, given I'm not a fan of really long names 😅 )

HendrikThePendric replied:
As discussed on Slack: we're keeping this for now because even if the naming is confusing/inaccurate, at least it is consistent with its sibling components. We can consider a rename for this entire dir in a separate PR

jenniferarnesen commented on src/components/DashboardsBar/DashboardsBar.js:
```diff
@@ -1,45 +1,15 @@
-            <InformationBlock />
+export const DashboardsBar = () => (
+    <div className={styles.toolbar} data-test="title-bar">
+        <div className={styles.blockCreationNavigation}>
```
Here the class name starts with "block", but on line 13, the component ends with "Block". Would it be more consistent to have this descriptor as a suffix?

HendrikThePendric replied:
We can do, but this was a conscious decision on my side and you'll see this pattern all across the code I write, so perhaps good to explain my motivation before I blindly change things around.... My reasoning is as follows:
- In JavaScript, in general uses `camelCase`. Classes, component and global constants are exceptions.
- When using CSS Modules in a React/Webpack application, you can import a default variable from the CSS file. 
- I see the `styles` / `classes` import as an object and each class-name as a property of that object, and the value of that object property is the classname-string incl. the unique ID.

I don't know if the last step is technically true but it seems to be an accurate enough representation of things. And thus, if `styles` is a "regular js object" and `styles.className` is a "regular js property" then is should always be written in camelCase.

jenniferarnesen replied:
It's not about the camelCase. It's about where the word `block` is in the name. In the css class, `block` is the prefix. In the component, it is the suffix.

HendrikThePendric replied:
Oh sorry I misread your comment 😊. I feel silly for writing such a lengthy reply that didn't make any sense. Note to self: spend more time reading to prevent wasting your own and other people's time 🤦 

Now that I have reread your comment, the suggestion makes perfect sense. Adding it to my todo list.

jenniferarnesen commented on src/components/Item/PrintTitlePageItem/Item.js:
```diff
@@ -3,7 +3,7 @@ import PropTypes from 'prop-types'
 import { connect } from 'react-redux'
 import { sGetIsEditing } from '../../../reducers/editDashboard.js'
-import { sGetNamedItemFilters } from '../../../reducers/itemFilters.js'
+import { msGetNamedItemFilters } from '../../../reducers/itemFilters.js'
```
What is the "ms" convention? I haven't seen it before.

HendrikThePendric replied:
True it is a new convention I suggested and introduced for "memoized selectors". I discussed it with Jan the other day in [this Slack thread](https://dhis2.slack.com/archives/G3TL0J145/p1738768600819609). Let me know it the convention and the motivation behind it makes sense to you. And also if we need to document this somewhere.

jenniferarnesen replied:
It's fine, just needed to know what it meant. Probably something we should share in some sort of conventions document or something.

jenniferarnesen commented on src/components/SystemSettingsProvider.js:
```diff
@@ -56,3 +60,15 @@ SystemSettingsProvider.propTypes = {
 export default SystemSettingsProvider
 
 export const useSystemSettings = () => useContext(SystemSettingsCtx)
+export const useHasSupersetSupport = () => {
```
`useSupersetIsSupported` perhaps?

HendrikThePendric replied:
Thanks for flagging this up. I think use `is` over `has` is preferable here as you suggest. In the end I went with `useIsSupersetSupported`, so that - when using this hook - a nice consistent `is`-prefixed boolean can be used:
```js
const isSupersetSupported = useIsSupersetSupported()
```

jenniferarnesen commented on src/modules/parseSupersetEmbeddedDashboardFieldValues.js:
```diff
@@ -0,0 +1,17 @@
+export const parseSupersetEmbeddedDashboardFieldValues = (values) => ({
```
Should `values` have a default {} ?

HendrikThePendric replied:
I think not. There is no valid use-case for calling this function without a `values` argument. I'd rather have this function crash than it silently producing unexpected results.

(Looking forward to us switching to TypeScript, which will allow us to differentiate between required and optional parameters. Obviously we could then still have debates on wether a parameter should be optional with a default, or required, but at least my intentions would be 100% clear to you when you read the code)

jenniferarnesen commented on src/modules/useSupersetEmbeddedDashboardFieldsState.js:
```diff
@@ -0,0 +1,94 @@
+import { useCallback, useReducer } from 'react'
+
+export const fieldNames = {
```
Some years ago we decided that exported constants should not be contained in an object since it can be error prone. For example if the consumer misspells the prop name, it can go undetected.

Suggest instead to export each of these as a separate constant.

jenniferarnesen commented on src/reducers/selected.js:
```diff
@@ -37,6 +40,43 @@ export const sGetSelected = (state) => state.selected
+}
+export const msGetSelectedSupersetEmbedData = createSelector(
+    [
+        (state) => state.selected?.embedded.id,
```
If `embedded` can be undefined (line 17), could this crash?

HendrikThePendric committed:
chore: update yarn.lock

HendrikThePendric committed:
chore: clarify supersetGateway api with comments

HendrikThePendric committed:
chore: rename hasSupersetSupport to isSupersetSupported

HendrikThePendric committed:
fix: correct position of conditional chain

HendrikThePendric committed:
chore: use individual constants for field names

jenniferarnesen commented on src/components/ConfigureSupersetEmbeddedDashboardModal/CreateSupersetEmbeddedDashboard.js:
```diff
@@ -0,0 +1,118 @@
+import { useDataMutation } from '@dhis2/app-runtime'
```
I'm thinking this component should be moved to a subcomponent of DashboardsBar. By convention that is how DashboardsBar components are organized. 

Also, I'm still hoping we can trim down this folder name and component names, and put "Modal" at the end of the Modal components (Create and Update)

jenniferarnesen commented on cypress.config.js:
```diff
@@ -5,10 +5,14 @@ const createEsbuildPlugin = require('@badeball/cypress-cucumber-preprocessor/esb
 const createBundler = require('@bahmutov/cypress-esbuild-preprocessor')
 const { chromeAllowXSiteCookies } = require('@dhis2/cypress-plugins')
 const { defineConfig } = require('cypress')
+// const {
```
We should just remove this commented code. At which point we need this, we can easily add it.

jenniferarnesen commented on cypress/e2e/embedded_superset_dashboard.cy.js:
```diff
@@ -0,0 +1,210 @@
+        cy.getByDataTest('dashboard-unstarred').should('be.visible')
+    })
+
+    it('can share the superset embedded dashboard', () => {
```
rename test to reflect that it is not actually sharing. i.e. "can open the sharing dialog"

jenniferarnesen commented on cypress/e2e/embedded_superset_dashboard.cy.js:
```diff
@@ -0,0 +1,210 @@
+            .and('contain', UUID)
+    })
+
+    it('has some options disabled in the action-bar', () => {
```
I think this test should be part of the previous test since it's just doing more confirmation that the dashboard was created. Also, I know we're already breaking Cypress best practice by having dependencies between tests, but the less of that we have the better.

jenniferarnesen commented on cypress/e2e/embedded_superset_dashboard.cy.js:
```diff
@@ -0,0 +1,210 @@
+        cy.get('.backdrop').should('be.visible').click()
+    })
+
+    it('shows the description', () => {
```
I'm a little concerned about changing the show description and letting it linger. I get what you're doing, but we also have another test suite for testing show/hide description. If these test suites run concurrently, they could interfere with each other.

HendrikThePendric replied:
Good point, I hadn't considered this.

jenniferarnesen commented on src/components/DashboardsBar/InformationBlock/LastUpdatedTag.js:
```diff
@@ -30,7 +30,7 @@ const LastUpdatedTag = ({ id }) => {
             {(props) => (
                 <div {...props}>
-                    <Tag maxWidth="400px">{message} </Tag>
+                    <Tag maxWidth="200px">{message} </Tag>
```
Is this changing the offline updated message? Won't 200px be too short?  
![image](https://github.com/user-attachments/assets/82e1a4f1-95f3-4cc5-a010-76732b63fc41)

HendrikThePendric committed:
chore: improve naming consistency for blocks

HendrikThePendric committed:
chore: solve sonarcube issue by removing commented code block

HendrikThePendric committed:
chore: reword e2e test description to better reflect its limitations

HendrikThePendric committed:
chore(e2e): merge disabled buttons step into creation step

cooper-joe committed:
fix: adjust external creation modal title

cooper-joe committed:
fix: adjust dashboard type radio components

cooper-joe committed:
fix: external creation code help text

HendrikThePendric committed:
chore: test showing and hiding the description in a single test step

HendrikThePendric committed:
chore: show description before asserting the updated description text

HendrikThePendric committed:
fix: restore last-updated-tag to original width

cooper-joe committed:
fix: labels in tests

jenniferarnesen commented on src/components/DashboardsBar/CreateDashboardButton/ChooseDashboardTypeModal.js:
```diff
@@ -51,14 +45,9 @@ export const ChooseDashboardTypeModal = ({
-                                }
-                            )}
+                            title={i18n.t('Internal', {
+                                nsSeparator: '###',
```
Is `nsSeparator` still needed (also lines 59-61)

HendrikThePendric committed:
chore: regenerate pot file after textual changes

HendrikThePendric committed:
fix: add max height of 180px to description field

HendrikThePendric committed:
fix: ensure Superset is capitalized in error message

HendrikThePendric committed:
fix: add tooltip around external data tag as in ui specs

HendrikThePendric committed:
chore: remove redundant nsSeparators in i18n.t calls

HendrikThePendric committed:
chore: rename files and and components according to PR feedback

Never use Superset and Embedded together, so that names are shorter and
all superset dashboards are embedded so the embedded part is redundant
anyway

Move and rename the superset dashboard  create/edit  components

HendrikThePendric committed:
fix: use logical css property

HendrikThePendric committed:
chore: fix formatting issue

HendrikThePendric committed:
fix: prevent flashing iframe error ui and show better error message

HendrikThePendric committed:
fix: adjust "External data" to "External source"

HendrikThePendric committed:
chore: update en.pot and yarn.lock

HendrikThePendric committed:
chore: adjust choose embedded dashboard type text

HendrikThePendric committed:
feat: omit tabbar for cascading sharing for embedded dashboards

HendrikThePendric committed:
fix: prevent flashing and show localised message based on error code

HendrikThePendric committed:
chore: use a separate message for E1005 error codes

HendrikThePendric committed:
chore: update pot file

HendrikThePendric committed:
docs: describe embedded dashboards feature

HendrikThePendric committed:
chore: add format script to package.json

HendrikThePendric committed:
chore: format docs markdown

HendrikThePendric committed:
fix: ensure confirm delete button gets focus and focus outline

HendrikThePendric committed:
fix: prevent modal content from getting focus outline

HendrikThePendric committed:
fix: make create and update superset modal forms scrollable

HendrikThePendric committed:
fix: show field error when superset embed id changes from valid to invalid

jenniferarnesen commented on src/components/DashboardsBar/ConfigureSupersetDashboard/styles/SupersetDashboardModal.module.css:
```diff
@@ -33,6 +33,10 @@
     outline-offset: -3px !important;
     text-decoration: underline !important;
 }
+/* Prevent focus outline from shwoing on modal content */
```
mispelling in comment

jenniferarnesen commented on i18n/en.pot:
```diff
@@ -5,14 +5,131 @@ msgstr ""
+msgid "External"
+msgstr "External"
+
+msgid "Embed a superset dashboard"
```
Should "superset" be capitalized?

jenniferarnesen commented on docs/dashboards.md:
```diff
@@ -58,6 +58,8 @@ In create/edit mode, you can add or change the dashboard title, description, das
 
 ![](resources/images/dashboard-create-mode.png)
 
+In some DHIS2 Core instances it is possible to either create a regular (internal) dashboard, or to embed a dashboard from an external source. See [this section](#creating-and-editing-an-external-dashboard) for more information about this.
```
I think this info should be emphasized as boxed text somehow, as an aside.

```
Note: if the system has been configured to support embedded dashboards, then you will have the option of creating a regular (internal) dashboard or embedding a dashboard from an external source. See [this section](#creating-and-editing-an-external-dashboard) for more information.
```

HendrikThePendric replied:
We discussed this on Slack and concluded we would use the sentence @jenniferarnesen suggested but without the `Note:` prefix, because in the DHIS2 Core docs Notes are typically in emphasised blocks and we do not want to emphasise this too much, since it is a feature which is not often available.

jenniferarnesen commented on docs/dashboards.md:
```diff
@@ -151,6 +153,27 @@ If you have access to delete the dashboard, then there will be a **Delete** butt
+-   Since the dashboard content is coming from an external system (Superset), there is no need to add dashboard items
+-   A "Superset Embed ID" needs to be provided so the app can retrieve the correct dashboard from the external Superset service
+-   The "Show chart controls on dashboard items" checkbox controls whether or not each embedded dashboard item has a dropdown menu
+-   The "Expand filters" checkbox controls whether or not the filter section of the embedded dashboard is expanded or collapsed
```
is initially expanded or collapsed

jenniferarnesen commented on docs/dashboards.md:
```diff
@@ -151,6 +153,27 @@ If you have access to delete the dashboard, then there will be a **Delete** butt
+
+![Superset dashboard with both checkboxes enabled](resources/images/superset-dashboard-with-annotations.png)
+
+Once the dashboard is created, you can star/unstar, edit and share it. You can also show and hide the description and close the dashboard. But it is not possible to show it in slideshow mode, filter it, or enter print view mode.
```
Remove the "But"

HendrikThePendric committed:
docs: adjust texts based on review comments

HendrikThePendric committed:
chore: fix i18n typo and update pot file

HendrikThePendric committed:
chore: fix typo in comment

HendrikThePendric committed:
fix: update superset embed id error visibility logic

HendrikThePendric committed:
fix: only show retry button for errors without error code

HendrikThePendric committed:
fix: ensure the first input gets focus after field values are reset

HendrikThePendric committed:
feat: add length validation for code field
